### ISSUE TITLE
Microsoft Windows Unquoted Service Path Enumeration (Nessus Plugin ID 63155)

### ISSUE THREAD

**Issue by mspncp** [mspncp/win32#1]
Sorry @drewbatgit for using this 'backdoor' to raise an issue for [MicrosoftDocs/win32](https://github.com/MicrosoftDocs/win32) :

_(well, not really sorry; why are issues still disabled?)_

For many years now, the Nessus Pentest tool has plagued uncounted companies and their software developers by reporting a [Microsoft Windows Unquoted Service Path Enumeration (Nessus Plugin ID 63155)](https://www.tenable.com/plugins/nessus/63155), because they forgot to add quotes around their service path.  As a consequence, due to the strange search rules of Windows, if it encounters an unquoted service path like

```
C:\Program Files\...\service.exe
```

it will attempt to execute `C:\Program` first, which will result in a privilege escalation if an adversary manages to place a malicious file `C:\Program.exe` into the root directory. In view of this fact, it surprises me that the official Microsoft example still advertises code which creates such unquoted entries:

https://github.com/mspncp/win32/blob/c47d327df5ac30685f2e84352d6cb91990b1887e/desktop-src/Services/svc-cpp.md?plain=1#L92-L96

https://github.com/mspncp/win32/blob/c47d327df5ac30685f2e84352d6cb91990b1887e/desktop-src/Services/svc-cpp.md?plain=1#L113-L126

I am aware that by default Microsoft only allows creation of subdirectories, not files, in the root directory. At least that's what I observe on a clean Windows 10 installation:

```
icacls C:\
C:\ BUILTIN\Administrators:(OI)(CI)(F)
    NT-AUTORITY\SYSTEM:(OI)(CI)(F)
    BUILTIN\Users:(OI)(CI)(RX)
    NT-AUTORITY\Authenticated Users:(OI)(CI)(IO)(M)
    NT-AUTORITY\Authenticated Users:(AD)
    Mandatory Label\High Integrity:(OI)(NP)(IO)(NW)
```
Here the corresponding SDDL String:
```
D:PAI(A;OICI;FA;;;BA)(A;OICI;FA;;;SY)(A;OICI;0x1200a9;;;BU)(A;OICIIO;SDGXGWGR;;;AU)(A;;LC;;;AU)S:P(ML;OINPIO;NW;;;HI)
```

So placing a malicious file `C:\Program.exe` into the root directory would not be possible without administrative privileges, unless somebody changed the default permissions carelessely.

Unfortunately, Nessus does not seem to know that, so it continuous torturing application developers. Actually, I would have preferred if Microsoft had fixed this problem properly long ago by making `CreateService()` more robust and having it add the quotes automatically if necessary. I was not able to report this to Tenable, because you need to be a customer to post issues on their community platform. So it would be great if you (Microsoft) could not only fix this example but also forward my report to your security team to have them fix the issue properly and permanently, and/or ask Tenable to make their Nessus test more sophisticated, i.e., check the permissions in the root directory as well.

Regards,

Matthias

**Comment by stevewhims** [comment]
Thanks @mspncp! We've updated the code example in the [Svc.cpp](https://docs.microsoft.com/windows/win32/services/svc-cpp) topic. That change is now live.

I don't have permissions to close this issue.

Many thanks!
-Steve

**Comment by mspncp** [comment]
Thank you, too, that was really a fast response!  Just for my curiosity: how long does it take (on average) until the changes propagate from GitHub to docs.microsoft.com?

https://docs.microsoft.com/en-us/windows/win32/services/svc-cpp

**Comment by drewbatgit** [comment]
For this content set, it should be less than an hour.

**Comment by mspncp** [comment]
@drewbatgit thanks for the information. It seems like the delay for the localized versions is larger? The [German version](https://docs.microsoft.com/de-de/windows/win32/services/svc-cpp) still shows the old code, while the english version was updated already yesterday.


@stevewhims on second thought I'm actually a little bit disappointed that you just added a belittling comment to the change:

![grafik](https://user-images.githubusercontent.com/11467854/160002690-75e4718b-9e79-4971-934d-161711612a3e.png)

IMHO  you missed a good chance to explain and warn users about the potential unquoted service path issue. (Maybe also mention that the "typical" case `C:\Program Files\...` is safe on current Windows versions.) That would probably be too lengthy for a comment in the code, but could be added as a caveat remark to the webpage.

Actually, a proper warning about the security implications would be even better placed in the API documentation for the [CreateService()](https://docs.microsoft.com/en-us/windows/win32/api/winsvc/nf-winsvc-createservicea) function. Although the documentation states that the path must be quoted, it does not mention its importance.

![grafik](https://user-images.githubusercontent.com/11467854/160006611-bd7a45e1-0f98-4368-9278-9b94e46359b7.png).

And last but not least: why doesn't Microsoft just fix the problem once and for all time by making the SCM smarter and having it add the quotes automatically if they are missing when starting a service?!

Would you mind forwarding my proposal(s) to the relevant department?

**Comment by stevewhims** [comment]
Thanks, @mspncp! I've updated the comment. I think the description from the CreateService function explains it better than my first attempt did. And I've forwarded this GitHub issue to the feature team owners so that they can see the feedback about SCM behavior (regarding auto-quoting where necessary). They'll also decide whether or not more emphasis needs to be placed on *why* the path needs quoting.

Thanks!
-Steve
